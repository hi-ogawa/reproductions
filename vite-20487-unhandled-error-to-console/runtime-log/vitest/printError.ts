import { stripVTControlCharacters } from "node:util";

export const c = {
  gray: (s: string) => s,
  red: (s: string) => s,
};

export function generateCodeFrame(
  source: string,
  indent = 0,
  loc: { line: number; column: number } | number,
  range = 2,
): string {
  const start =
    typeof loc === "object"
      ? positionToOffset(source, loc.line, loc.column)
      : loc;
  const end = start;
  const lines = source.split(lineSplitRE);
  const nl = /\r\n/.test(source) ? 2 : 1;
  let count = 0;
  let res: string[] = [];

  const columns = process.stdout?.columns || 80;

  for (let i = 0; i < lines.length; i++) {
    count += lines[i].length + nl;
    if (count >= start) {
      for (let j = i - range; j <= i + range || end > count; j++) {
        if (j < 0 || j >= lines.length) {
          continue;
        }

        const lineLength = lines[j].length;
        const strippedContent = stripVTControlCharacters(lines[j]);

        if (strippedContent.startsWith("//# sourceMappingURL")) {
          continue;
        }

        // too long, maybe it's a minified file, skip for codeframe
        if (strippedContent.length > 200) {
          return "";
        }

        res.push(
          lineNo(j + 1) +
            truncateString(lines[j].replace(/\t/g, " "), columns - 5 - indent),
        );

        if (j === i) {
          // push underline
          const pad = start - (count - lineLength) + (nl - 1);
          const length = Math.max(
            1,
            end > count ? lineLength - pad : end - start,
          );
          res.push(lineNo() + " ".repeat(pad) + c.red("^".repeat(length)));
        } else if (j > i) {
          if (end > count) {
            const length = Math.max(1, Math.min(end - count, lineLength));
            res.push(lineNo() + c.red("^".repeat(length)));
          }
          count += lineLength + 1;
        }
      }
      break;
    }
  }

  if (indent) {
    res = res.map((line) => " ".repeat(indent) + line);
  }

  return res.join("\n");
}

function lineNo(no: number | string = "") {
  return c.gray(`${String(no).padStart(3, " ")}| `);
}

export const lineSplitRE: RegExp = /\r?\n/;

export function positionToOffset(
  source: string,
  lineNumber: number,
  columnNumber: number,
): number {
  const lines = source.split(lineSplitRE);
  const nl = /\r\n/.test(source) ? 2 : 1;
  let start = 0;

  if (lineNumber > lines.length) {
    return source.length;
  }

  for (let i = 0; i < lineNumber - 1; i++) {
    start += lines[i].length + nl;
  }

  return start + columnNumber;
}

export function offsetToLineNumber(source: string, offset: number): number {
  if (offset > source.length) {
    throw new Error(
      `offset is longer than source length! offset ${offset} > length ${source.length}`,
    );
  }
  const lines = source.split(lineSplitRE);
  const nl = /\r\n/.test(source) ? 2 : 1;
  let counted = 0;
  let line = 0;
  for (; line < lines.length; line++) {
    const lineLength = lines[line].length + nl;
    if (counted + lineLength >= offset) {
      break;
    }

    counted += lineLength;
  }
  return line + 1;
}

export function truncateString(text: string, maxLength: number): string {
  const plainText = stripVTControlCharacters(text);

  if (plainText.length <= maxLength) {
    return text;
  }

  return `${plainText.slice(0, maxLength - 1)}â€¦`;
}
