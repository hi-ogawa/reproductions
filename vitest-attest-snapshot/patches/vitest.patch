diff --git a/dist/chunks/vi.JMQoNY_Z.js b/dist/chunks/vi.JMQoNY_Z.js
index e75ae84a3fe80dd116f8c599deb483958b843c17..04104f9a7b8577543570eba58539bd35939085ab 100644
--- a/dist/chunks/vi.JMQoNY_Z.js
+++ b/dist/chunks/vi.JMQoNY_Z.js
@@ -194,7 +194,7 @@ function requireChaiSubset () {
 				}
 			});
 
-		}).call(chaiSubset); 
+		}).call(chaiSubset);
 	} (chaiSubset$1));
 	return chaiSubset$1.exports;
 }
@@ -322,6 +322,18 @@ const SnapshotPlugin = (chai, utils) => {
         );
       }
       const expected = utils.flag(this, "object");
+      if (globalThis.process?.env?.ATTEST_skipTypes) {
+        const val = expected
+        if (
+          !!val &&
+          typeof val === 'object' &&
+          'constructor' in val &&
+          typeof val.constructor === 'function' &&
+          val.constructor.name === 'ChainableAssertions'
+        ) {
+          return
+        }
+      }
       const error = utils.flag(this, "error");
       if (typeof properties === "string") {
         message = inlineSnapshot;
@@ -810,7 +822,7 @@ function requireDeprecated () {
 		    } else {
 		        console.log(msg);
 		    }
-		}; 
+		};
 	} (deprecated));
 	return deprecated;
 }
@@ -1371,7 +1383,7 @@ function requireTypeDetect () {
 
 		return typeDetect;
 
-		}))); 
+		})));
 	} (typeDetect$1));
 	return typeDetect$1.exports;
 }
