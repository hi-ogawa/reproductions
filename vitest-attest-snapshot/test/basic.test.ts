import { test, expect, type Assertion } from "vitest";
import { attest, caller, getTypeAssertionsAtPosition } from "@ark/attest";

//
// via custom vitest snapshot seiralizer
// Q. easier way to detect ChainableAssertions and proxy etc..?

// for attest().type.toString and type.errors
expect.addSnapshotSerializer({
  test(val: unknown) {
    return (
      !!val &&
      typeof val === "function" &&
      typeof (val as any).snap === "function"
    );
  },
  serialize(val, _config, _indentation, _depth, _refs, _printer) {
    return val.serializedActual;
  },
});

// for attest() to work like attest().type.toString
expect.addSnapshotSerializer({
  test(val: unknown) {
    return (
      !!val &&
      typeof val === "object" &&
      "snap" in val &&
      typeof val.snap === "function"
    );
  },
  serialize(val, _config, _indentation, _depth, _refs, _printer) {
    return val.type.toString.serializedActual;
  },
});

test("snap", (ctx) => {
  if (process.env["ATTEST_skipTypes"]) {
    ctx.skip();
  }

  expect(attest(1 + 2).type.toString).toMatchInlineSnapshot(`number`);
  expect(attest("1" + 2).type.toString).toMatchInlineSnapshot(`string`);
  expect(
    attest(
      () =>
        // prettier-ignore
        // @ts-expect-error
        "1" / 2,
    ).type.errors,
  ).toMatchInlineSnapshot(
    `The left-hand side of an arithmetic operation must be of type 'any', 'number', 'bigint' or an enum type.`,
  );

  // make it work like .type.toString by default
  expect(attest(1 + 2)).toMatchInlineSnapshot(`number`);
  expect(attest("1" + 2)).toMatchInlineSnapshot(`string`);

  // Q. easier to get `completions`?
  function getCompletions(t: any) {
    t.completions; // getter side effect seems to do magic
    return t.serializedActual;
  }
  expect(
    getCompletions(
      attest(
        () =>
          // prettier-ignore
          // @ts-expect-error
          (1 + 2)["to"],
      ),
    ),
  ).toMatchInlineSnapshot(`
    {
      "to": [
        "toExponential",
        "toFixed",
        "toLocaleString",
        "toPrecision",
        "toString",
      ],
    }
  `);
});

//
// via custom attest alias
//
function expectType(_actualValue: unknown): Assertion<any> {
  if (process.env["ATTEST_skipTypes"]) {
    // return no-op proxy
    return new Proxy(
      {},
      {
        get: () => () => {},
      },
    ) as any;
  }
  const position = caller();
  const types = getTypeAssertionsAtPosition(position);
  return expect(types[0][1].args[0].type);
}

test("custom attest alias", () => {
  expectType((1 + 2).toFixed).toMatchInlineSnapshot(
    `"(fractionDigits?: number | undefined) => string"`,
  );
  expectType((1 + 3).toPrecision).toMatchSnapshot();
});
