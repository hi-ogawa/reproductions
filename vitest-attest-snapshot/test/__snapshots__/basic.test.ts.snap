// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`custom attest alias 2`] = `"(precision?: number | undefined) => string"`;
